sdks/python/apache_beam/transforms/enrichment.py

```diff
@@ -0,0 +1,130 @@
+from apache_beam.io.requestresponse import Caller
+from apache_beam.io.requestresponse import PreCallThrottler
+from apache_beam.io.requestresponse import Repeater
+from apache_beam.io.requestresponse import RequestResponseIO
+from apache_beam.io.requestresponse import ShouldBackOff
+
+__all__ = [
+    "EnrichmentSourceHandler",
+    "Enrichment",
+]
+
+InputT = TypeVar('InputT')
+OutputT = TypeVar('OutputT')
+
+
+def cross_join(element):
+  """cross_join performs a cross join between two `beam.Row` objects.
+
+    Joins the columns of the right `beam.Row` onto the left `beam.Row`.
+
+    Args:
+      element (Tuple): A tuple containing two beam.Row objects -
+        request and response.
+
+    Returns:
+      beam.Row: `beam.Row` containing the merged columns.
+  """
+  right_dict = element[1].as_dict()
+  left_dict = element[0].as_dict()
+  for k, v in right_dict.items():
+    left_dict[k] = v
+  return beam.Row(**left_dict)
+
+
+class EnrichmentSourceHandler(Caller[InputT, OutputT]):
+  """Wrapper class for :class:`apache_beam.io.requestresponse.Caller`.
+
+  Ensure that the implementation of ``__call__`` method returns a tuple
+  of `beam.Row`  objects.
+  """
+
+  pass
+
+
+class Enrichment(beam.PTransform[beam.PCollection[InputT],
+                                 beam.PCollection[OutputT]],
+                 Generic[InputT, OutputT]):
+  """A :class:`apache_beam.transforms.enrichment.Enrichment` transform to
+  enrich elements in a PCollection.
+
+  Uses the :class:`apache_beam.transforms.enrichment.EnrichmentSourceHandler`
+  to enrich elements by joining the metadata from external source.
+
+  Processes an input :class:`~apache_beam.pvalue.PCollection` of `beam.Row` by
+  applying a :class:`apache_beam.transforms.enrichment.EnrichmentSourceHandler`
+  to each element and returning the enriched
+  :class:`~apache_beam.pvalue.PCollection`.
+
+  Args:
+    source_handler: Handles source lookup and metadata retrieval.
+      Implements the
+      :class:`apache_beam.transforms.enrichment.EnrichmentSourceHandler`
+    join_fn: A lambda function to join original element with lookup metadata.
+      Defaults to `CROSS_JOIN`.
+    timeout: (Optional) timeout for source requests. Defaults to 30 seconds.
+    should_backoff: (Optional) backoff strategy function.
+    repeater: (Optional) retry Repeater.
+    cache_reader: (Optional) CacheReader for reading cache.
+    cache_writer: (Optional) CacheWriter for writing cache.
+    throttler: (Optional) Throttler mechanism to throttle source requests.
+  """
+  def __init__(
+      self,
+      source_handler: EnrichmentSourceHandler,
+      join_fn: Callable = cross_join,
+      timeout: Optional[float] = DEFAULT_TIMEOUT_SECS,
+      should_backoff: Optional[ShouldBackOff] = None,
+      repeater: Optional[Repeater] = None,
+      cache_reader: Optional[CacheReader] = None,
+      cache_writer: Optional[CacheWriter] = None,
+      throttler: Optional[PreCallThrottler] = None):
```

Comment:
These params are not functional at this point but I expect them to be before 2.54 cut. So I left it here. I'm open to adding them as param once they are functional. No preference here!

Reply:
Lets remove them for now. That way if something comes up (sickness, other priorities, etc...) this doesn't get accidentally left in

Reply:
sounds good!